File: domain/src/main/java/com/arturogutierrez/openticator/domain/category/CategorySelector.java
package com.arturogutierrez.openticator.domain.category;

import com.arturogutierrez.openticator.domain.category.model.Category;
import com.arturogutierrez.openticator.domain.category.model.EmptyCategory;
import rx.Observable;
import rx.subjects.PublishSubject;

public class CategorySelector {

  private final PublishSubject<Category> publishSubject;
  private Category currentCategory;

  public CategorySelector() {
    this.currentCategory = new EmptyCategory();
    this.publishSubject = PublishSubject.create();
  }

  public Observable<Category> getSelectedCategory() {
    return publishSubject.startWith(currentCategory);
  }

  public void setSelectedCategory(Category category) {
    currentCategory = category;
    publishSubject.onNext(category);
  }

  public void removeSelectedCategory() {
    setSelectedCategory(new EmptyCategory());
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/category/CategorySelector.kt
package com.arturogutierrez.openticator.domain.category

import com.arturogutierrez.openticator.domain.category.model.Category
import io.reactivex.Flowable
import io.reactivex.processors.BehaviorProcessor
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CategorySelector @Inject constructor() {

  private val publishSubject = BehaviorProcessor.create<Category>()
  private var currentCategory = Category.empty

  val selectedCategory: Flowable<Category>
    get() = publishSubject.startWith(currentCategory)

  fun setSelectedCategory(category: Category) {
    currentCategory = category
    publishSubject.onNext(category)
  }

  fun removeSelectedCategory() {
    setSelectedCategory(Category.empty)
  }
}
